File: scripts/analyze_agent_rules.mjs
#!/usr/bin/env node
/**
 * Agent Rules Analyzer — Browser Automation (Node.js)
 * 
 * Automates submission of config files to agentrulegen.com/analyze
 * since no API exists.
 * 
 * Requirements: npx playwright (already installed)
 * 
 * Usage:
 *   node analyze_agent_rules.mjs .ruler/AGENTS.md
 *   node analyze_agent_rules.mjs .claude/hooks.yaml rulesync.jsonc
 */

import { chromium } from 'playwright';
import fs from 'fs';
import path from 'path';

const ANALYZE_URL = 'https://agentrulegen.com/analyze';
const MAX_CHARS = 10000;

async function analyzeRules(filePath) {
    const content = fs.readFileSync(filePath, 'utf-8');
    const truncatedContent = content.length > MAX_CHARS 
        ? content.slice(0, MAX_CHARS) 
        : content;
    
    console.log(`Analyzing: ${filePath} (${truncatedContent.length} chars)`);
    
    const browser = await chromium.launch({ headless: true });
    const page = await browser.newPage();
    
    try {
        // Navigate
        console.log('Navigating to agentrulegen.com/analyze...');
        await page.goto(ANALYZE_URL, { waitUntil: 'networkidle' });
        
        // Wait for textarea
        const textarea = await page.waitForSelector('textarea', { timeout: 10000 });
        
        // Fill content
        console.log('Submitting content...');
        await textarea.fill(truncatedContent);
        
        // Click analyze button
        const analyzeBtn = await page.waitForSelector('button:has-text("Analyze")', { timeout: 5000 });
        await analyzeBtn.click();
        
        // Wait for results
        console.log('Waiting for analysis results...');
        await page.waitForSelector(':text("results"), :text("redundant"), :text("essential")', { timeout: 60000 });
        
        // Scrape results
        console.log('Scraping results...');
        const bodyText = await page.innerHTML('body');
        
        // Try to extract structured data
        const results = {
            file: filePath,
            charCount: truncatedContent.length,
            rawResponse: bodyText,
            redundant: [],
            essential: [],
            improvable: [],
            missing: []
        };
        
        // Extract categories (look for common patterns)
        try {
            const redundantText = await page.textContent(':text-matches("redundant", "i") ~ ul, :text-matches("redundant", "i") + *');
            if (redundantText) results.redundant = redundantText.split('\n').filter(Boolean);
        } catch {}
        
        try {
            const essentialText = await page.textContent(':text-matches("essential", "i") ~ ul, :text-matches("essential", "i") + *');
            if (essentialText) results.essential = essentialText.split('\n').filter(Boolean);
        } catch {}
        
        try {
            const improvableText = await page.textContent(':text-matches("improvable", "i") ~ ul, :text-matches("improvable", "i") + *');
            if (improvableText) results.improvable = improvableText.split('\n').filter(Boolean);
        } catch {}
        
        return results;
        
    } catch (error) {
        return {
            file: filePath,
            error: error.message,
            success: false
        };
    } finally {
        await browser.close();
    }
}

async function batchAnalyze(files) {
    const results = [];
    for (const file of files) {
        const result = await analyzeRules(file);
        results.push(result);
        console.log(`Complete: ${file}`);
        // Delay between requests
        await new Promise(r => setTimeout(r, 2000));
    }
    return results;
}

// Main
const files = process.argv.slice(2);

if (files.length === 0) {
    console.log('Usage: node analyze_agent_rules.mjs <file> [file2] [file3]');
    console.log('Example: node analyze_agent_rules.mjs .ruler/AGENTS.md .claude/hooks.yaml');
    process.exit(1);
}

// Validate files exist
for (const f of files) {
    if (!fs.existsSync(f)) {
        console.log(`ERROR: File not found: ${f}`);
        process.exit(1);
    }
}

// Run
console.log('Starting analysis...\n');
batchAnalyze(files).then(results => {
    console.log('\n=== ANALYSIS RESULTS ===');
    console.log(JSON.stringify(results, null, 2));
    
    // Save to file
    const outputFile = 'agent_rules_analysis.json';
    fs.writeFileSync(outputFile, JSON.stringify(results, null, 2));
    console.log(`\nSaved to: ${outputFile}`);
});

File: scripts/analyze_agent_rules.py
#!/usr/bin/env python3
"""
Agent Rules Analyzer — Browser Automation

Automates submission of config files to agentrulegen.com/analyze
since no API exists.

Requirements:
- playwright (npm install -g playwright)
- Or: npx playwright python

Usage:
    python3 analyze_agent_rules.py .ruler/AGENTS.md
    python3 analyze_agent_rules.py .claude/hooks.yaml
    python3 analyze_agent_rules.py rulesync.jsonc
"""

import sys
import json
import asyncio
from pathlib import Path

try:
    from playwright.async_api import async_playwright
except ImportError:
    print("ERROR: playwright not installed")
    print("Install: pip install playwright && playwright install chromium")
    sys.exit(1)

ANALYZE_URL = "https://agentrulegen.com/analyze"
MAX_CHARS = 10000


async def analyze_rules(file_path: str) -> dict:
    """
    Submit file content to agentrulegen.com/analyze and scrape results.
    
    Args:
        file_path: Path to config file (.ruler/AGENTS.md, hooks.yaml, etc.)
    
    Returns:
        dict with keys: redundant, essential, improvable, missing
    """
    # Read file content
    content = Path(file_path).read_text()
    
    if len(content) > MAX_CHARS:
        print(f"WARNING: Content exceeds {MAX_CHARS} chars ({len(content)}), truncating")
        content = content[:MAX_CHARS]
    
    print(f"Analyzing: {file_path} ({len(content)} chars)")
    
    async with async_playwright() as p:
        # Launch browser
        browser = await p.chromium.launch(headless=True)
        page = await browser.new_page()
        
        try:
            # Navigate to analyze page
            print("Navigating to agentrulegen.com/analyze...")
            await page.goto(ANALYZE_URL, wait_until="networkidle")
            
            # Wait for text area to be visible
            text_area = await page.wait_for_selector("textarea", timeout=10000)
            
            # Fill in content
            print("Submitting content...")
            await text_area.fill(content)
            
            # Click analyze button (look for button with "Analyze" text)
            analyze_btn = await page.wait_for_selector("button:has-text('Analyze')", timeout=5000)
            await analyze_btn.click()
            
            # Wait for results (look for results container)
            print("Waiting for analysis results...")
            await page.wait_for_selector("[class*='result'], [class*='analysis'], :text('redundant'), :text('essential')", timeout=60000)
            
            # Scrape results
            print("Scraping results...")
            results_text = await page.inner_text("body")
            
            # Try to find structured results
            results = {
                "file": file_path,
                "char_count": len(content),
                "raw_response": results_text,
                "redundant": [],
                "essential": [],
                "improvable": [],
                "missing": []
            }
            
            # Parse common patterns in response
            # "Redundant rules:" section
            redundant_section = await page.query_selector(":text('Redundant') ~ *")
            if redundant_section:
                results["redundant"] = (await redundant_section.inner_text()).split("\n")
            
            essential_section = await page.query_selector(":text('Essential') ~ *")
            if essential_section:
                results["essential"] = (await essential_section.inner_text()).split("\n")
            
            improvable_section = await page.query_selector(":text('Improvable') ~ *")
            if improvable_section:
                results["improvable"] = (await improvable_section.inner_text()).split("\n")
            
            return results
            
        except Exception as e:
            return {
                "file": file_path,
                "error": str(e),
                "success": False
            }
        finally:
            await browser.close()


async def batch_analyze(files: list[str]) -> list[dict]:
    """Analyze multiple config files."""
    results = []
    for file_path in files:
        result = await analyze_rules(file_path)
        results.append(result)
        print(f"Complete: {file_path}")
        # Small delay between requests to avoid rate limiting
        await asyncio.sleep(2)
    return results


def main():
    if len(sys.argv) < 2:
        print("Usage: python3 analyze_agent_rules.py <file> [file2] [file3]")
        print("Example: python3 analyze_agent_rules.py .ruler/AGENTS.md .claude/hooks.yaml")
        sys.exit(1)
    
    files = sys.argv[1:]
    
    # Validate files exist
    for f in files:
        if not Path(f).exists():
            print(f"ERROR: File not found: {f}")
            sys.exit(1)
    
    # Run analysis
    results = asyncio.run(batch_analyze(files))
    
    # Output JSON
    print("\n=== ANALYSIS RESULTS ===")
    print(json.dumps(results, indent=2))
    
    # Save to file
    output_file = "agent_rules_analysis.json"
    Path(output_file).write_text(json.dumps(results, indent=2))
    print(f"\nSaved to: {output_file}")


if __name__ == "__main__":
    main()
